client/tracing/src/logging/mod.rs

```diff
@@ -0,0 +1,349 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate logging library.
+//!
+//! This crate uses tokio's [tracing](https://github.com/tokio-rs/tracing/) library for logging.
+
+#![warn(missing_docs)]
+
+mod event_format;
+mod layers;
+
+pub use sc_tracing_proc_macro::*;
+
+use tracing::Subscriber;
+use tracing_subscriber::{
+	filter::Directive, fmt::time::ChronoLocal, layer::SubscriberExt, registry::LookupSpan,
+	FmtSubscriber, Layer,
+};
+
+pub use event_format::*;
+pub use layers::*;
+
+/// Get a new default tracing's `Subscriber` and a sc-telemetry's `Telemetries` objects.
+///
+/// When running in a browser, the `telemetry_external_transport` should be provided.
+pub fn get_default_subscriber_and_telemetries(
+	pattern: &str,
+	telemetry_external_transport: Option<sc_telemetry::ExtTransport>,
+) -> std::result::Result<
+	(
+		impl Subscriber + for<'a> LookupSpan<'a>,
+		sc_telemetry::Telemetries,
+	),
+	String,
+> {
+	get_default_subscriber_and_telemetries_internal(
+		parse_directives(pattern),
+		telemetry_external_transport,
+	)
+}
+
+/// Get a new default tracing's `Subscriber` and a sc-telemetry's `Telemetries` objects with
+/// profiling enabled.
+///
+/// When running in a browser, the `telemetry_external_transport` should be provided.
+pub fn get_default_subscriber_and_telemetries_with_profiling(
+	pattern: &str,
+	telemetry_external_transport: Option<sc_telemetry::ExtTransport>,
+	tracing_receiver: crate::TracingReceiver,
+	profiling_targets: &str,
+) -> std::result::Result<
+	(
+		impl Subscriber + for<'a> LookupSpan<'a>,
+		sc_telemetry::Telemetries,
+	),
+	String,
+> {
+	let (subscriber, telemetries) = get_default_subscriber_and_telemetries_internal(
+		parse_directives(pattern)
+			.into_iter()
+			.chain(parse_directives(profiling_targets).into_iter()),
+		telemetry_external_transport,
+	)?;
+	let profiling = crate::ProfilingLayer::new(tracing_receiver, profiling_targets);
+
+	Ok((subscriber.with(profiling), telemetries))
+}
+
+// Common implementation for `get_default_subscriber_and_telemetries` and
+// `get_default_subscriber_and_telemetries_with_profiling`.
+fn get_default_subscriber_and_telemetries_internal(
+	extra_directives: impl IntoIterator<Item = Directive>,
+	telemetry_external_transport: Option<sc_telemetry::ExtTransport>,
+) -> std::result::Result<
+	(
+		impl Subscriber + for<'a> LookupSpan<'a>,
+		sc_telemetry::Telemetries,
+	),
+	String,
+> {
+	if let Err(e) = tracing_log::LogTracer::init() {
```

Comment:
Will this fail for Cumulus? 

Ahh or do we don't call this 2 times?

Reply:
You only call this once yes. (it's related to global logger initialization)